x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/CombineDisjunctions.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.optimizer.rules;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.expression.Literal;
+import org.elasticsearch.xpack.esql.core.expression.predicate.logical.Or;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.expression.function.scalar.ip.CIDRMatch;
+import org.elasticsearch.xpack.esql.expression.predicate.operator.comparison.Equals;
+import org.elasticsearch.xpack.esql.expression.predicate.operator.comparison.In;
+
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.elasticsearch.xpack.esql.core.expression.predicate.Predicates.combineOr;
+import static org.elasticsearch.xpack.esql.core.expression.predicate.Predicates.splitOr;
+import static org.elasticsearch.xpack.esql.type.EsqlDataTypeConverter.ipToString;
+
+/**
+ * Combine disjunctive Equals or In expression on the same field into an In expression.
+ * This rule looks for both simple equalities:
+ * 1. a == 1 OR a == 2 becomes a IN (1, 2)
+ * and combinations of In
+ * 2. a == 1 OR a IN (2) becomes a IN (1, 2)
+ * 3. a IN (1) OR a IN (2) becomes a IN (1, 2)
+ * Combine disjunctive Equals, In or CIDRMatch expressions on the same field into a CIDRMatch expression.
+ * 4. CIDRMatch(a, ip1) OR CIDRMatch(a, ip2) OR a = ip3 or a IN (ip4, ip5) becomes CIDRMatch(a, ip1, ip2, ip3, ip4, ip5)
+ * <p>
+ * This rule does NOT check for type compatibility as that phase has been
+ * already be verified in the analyzer.
+ */
+public final class CombineDisjunctions extends OptimizerRules.OptimizerExpressionRule<Or> {
+    public CombineDisjunctions() {
+        super(OptimizerRules.TransformDirection.UP);
+    }
+
+    protected static In createIn(Expression key, List<Expression> values, ZoneId zoneId) {
+        return new In(key.source(), key, values);
+    }
+
+    protected static Equals createEquals(Expression k, Set<Expression> v, ZoneId finalZoneId) {
+        return new Equals(k.source(), k, v.iterator().next(), finalZoneId);
+    }
+
+    protected static CIDRMatch createCIDRMatch(Expression k, List<Expression> v) {
+        return new CIDRMatch(k.source(), k, v);
+    }
+
+    @Override
+    public Expression rule(Or or) {
+        Expression e = or;
+        // look only at equals, In and CIDRMatch
+        List<Expression> exps = splitOr(e);
+
+        Map<Expression, Set<Expression>> ins = new LinkedHashMap<>();
```

Comment:
Does it matter if it's Linked map or a simple map can do it? Does it have to do with the deterministic way in which the query dsl is being built?

Reply:
> Does it matter if it's Linked map or a simple map can do it? Does it have to do with the deterministic way in which the query dsl is being built?

A simple map is fine I think.